File: leetcode/3-longest-substring-without-repeating-characters.py
# https://leetcode.com/problems/longest-substring-without-repeating-characters/ (top-interview-questions)

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        start, longest = 0, 0
        charSet = set()
        for end in range(len(s)):
            while s[end] in charSet:
                charSet.remove(s[start])
                start += 1
            charSet.add(s[end])
            longest = max(end - start + 1, longest)
        return longest

File: leetcode/3-longest-substring-without-repeating-characters.js
// https://leetcode.com/problems/longest-substring-without-repeating-characters/ (top-interview-questions)

/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
  const charSet = new Set();
  let length = 0; // length of the longest substring

  let i = 0;
  for (let j = 0; j < s.length; j++) {
    while(charSet.has(s[j])) charSet.delete(s[i++]);

    charSet.add(s[j]);
    if (charSet.size > length) length = charSet.size;
  }

  return length;
};